### ISSUE TITLE
A11y initial focus on "dismiss barrier" when showing a dismissable AlertDialog 

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#46625]
Internal bug: b/145715671

Repro:
- Turn on Talkback.
- Create a barrier dismissable dialog via Flutter on iOS 13.2.

Note that:
- The initial focus is on entire screen (or in some cases on the system clock). 
- Unable to move the focus to the dialog by using swipes.
- Tapping into the dialog moves the focus as expected.

Expected Behavior:
- Should be able to swipe and highlight the dialog elements.
- If the dialog is not dismissable, a11y correctly focuses on the elements in the dialog first. We would expect the same behavior from a dismissable dialog. Initial focus should be on the dialog title. Swiping should cycle through the elements including dialog buttons and "screen" (dismiss action).

**Comment by mehmetf** [comment]
The workaround is to define the dialog as non-dismissable so this is not a blocker but it is still critical as this requires apps to change the dialog behavior.

**Comment by HansMuller** [comment]
CC @darrenaustin

**Comment by darrenaustin** [comment]
I am able repro this with:

```dart
import 'package:flutter/material.dart';

void main() => runApp(A11yDialogApp());

class A11yDialogApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Builder(
        builder: (BuildContext context) {
          return Scaffold(
            appBar: AppBar(title: const Text('Accessibility Dialog Test')),
            body: Center(
              child: RaisedButton(
                child: const Text('Dialog for me!'),
                onPressed: () {
                  showDialog<void>(
                    context: context,
//                    barrierDismissible: false,
                    builder: (BuildContext context) {
                      return AlertDialog(
                        content: const Text('Do we have focus?'),
                        actions: <Widget>[
                          FlatButton(
                            child: const Text('OK'),
                            onPressed: () { Navigator.pop(context); },
                          ),
                        ],
                      );
                    }
                  );
                },
              ),
            ),
          );
        },
      ),
    );
  }
}
```

It does appear to be a problem with the dismiss barrier getting the a11y focus instead of the dialog. Not sure why yet. Looking.

**Comment by darrenaustin** [comment]
I have an attempt to fix this still be reviewed and discussed. Sorry we didn't get this sorted out before the break.

**Comment by darrenaustin** [comment]
The core problem here is that TalkBack chooses to focus on the "dismiss barrier" instead of something in the dialog proper (not sure why, although maybe because it comes above the dialog in the semantic tree).  I attempted to fix this with #14513. However, it forces a behavior that may not be correct either (it always selects the dialog title). Another approach to this specific problem may be to just exclude the barrier from semantics altogether similar to the way the drawer was handled in https://github.com/flutter/flutter/pull/48841. Either way, I will let Chris take it from here as he is looking into the a11y engine bugs.

**Comment by mehmetf** [comment]
Customer is requesting an update on this. Please let us know if there's any progress. Thanks!

**Comment by HansMuller** [comment]
CC @cbracken

**Comment by mehmetf** [comment]
> May be to just exclude the barrier from semantics altogether similar to the way the drawer was handled in #48841.

I think that might be a legitimate workaround except that now accessible users won't be able to interact with the barrier to dismiss the dialog. I believe some UX patterns depend on that. If it is OK for drawer, perhaps it is OK for dialogs too?

**Comment by darrenaustin** [comment]
Perhaps, but the problem is that in general there is no consistent way to dismiss a dialog other than the barrier (or perhaps the back button on Android).  The dialog is not guaranteed to have an action button to dismiss it. So if we don't expose the barrier to semantics, there may not be a way for the user to dismiss the dialog via a11y.

**Comment by chunhtai** [comment]
Interestingly, the barrier dismissal is not focusable in android using accessibility focus. Not sure if it is a bug or intended behavior. however, it is ok for android because it has a backbutton. Not sure if the same thing is acceptable in IOS
